RecoJets/JetProducers/plugins/BetaStarVarProducer.cc

```diff
@@ -69,89 +72,103 @@ template <typename T>
-  std::vector<float> chargedHadronCHSEnergyFraction(nJet,-1);
+  // Create an injective mapping from jets to charged PF candidates with fromPV==0 within the jet cone.
+  // These are the PF candidates supposedly removed by the CHS method - i.e. removed charged pileup.
+  std::vector< double > jet2pue(nJet,0);
+  for ( unsigned int icand = 0; icand < srcPF->size(); ++icand ) {
+    auto cand = srcPF->ptrAt(icand);
+    if (cand->charge()==0 or cand->fromPV()!=0) continue;
+    // Looking for the closest match for this charged frompv==0 candidate
+    float dRMin = 999.;
+    int bestjet = -1;
+    for ( auto ijet=srcJet->begin(); ijet!=srcJet->end(); ++ijet ) {
+      float dR = reco::deltaR(*ijet,*cand);
+      if (dR<dRMin) {
+        dRMin = dR;
+        bestjet = ijet-srcJet->begin();
+      }
+    }
+    // If the candidate is closer than the jet radius to the jet axis, this is a PU particle of interest for the selected jet
+    if (dRMin<maxDR_) jet2pue[bestjet] += cand->energy();
+  }
+
+  std::vector<float> chargedFromPV0EnergyFraction(nJet,-1);
+  std::vector<float> chargedFromPV1EnergyFraction(nJet,-1);
+  std::vector<float> chargedFromPV2EnergyFraction(nJet,-1);
+  std::vector<float> chargedFromPV3EnergyFraction(nJet,-1);
 
   for ( unsigned int ij = 0; ij < nJet; ++ij ) {
-    auto jet = srcJet->ptrAt(ij);
-    std::tuple<float,float> vals = calculateCHSEnergies( jet, *srcPF );
-    auto chpuf = std::get<0>(vals);
-    auto chef  = std::get<1>(vals);
-    chargedHadronPUEnergyFraction[ij] = chpuf;
-    chargedHadronCHSEnergyFraction[ij] = chef;
+    auto ijet = srcJet->ptrAt(ij);
+    auto vals = calculateCHSEnergies( ijet, jet2pue[ij] );
+    auto chpuf0 = std::get<0>(vals);
+    auto chpuf1 = std::get<1>(vals);
+    auto chpuf2 = std::get<2>(vals);
+    auto chpuf3 = std::get<3>(vals);
+    //auto chef   = std::get<2>(vals);
+    chargedFromPV0EnergyFraction[ij] = chpuf0;
+    chargedFromPV1EnergyFraction[ij] = chpuf1;
+    chargedFromPV2EnergyFraction[ij] = chpuf2;
+    chargedFromPV3EnergyFraction[ij] = chpuf3;
   }
 
-  std::unique_ptr<edm::ValueMap<float>> chargedHadronPUEnergyFractionV(new edm::ValueMap<float>());
-  edm::ValueMap<float>::Filler fillerPU(*chargedHadronPUEnergyFractionV);
-  fillerPU.insert(srcJet,chargedHadronPUEnergyFraction.begin(),chargedHadronPUEnergyFraction.end());
-  fillerPU.fill();
-  iEvent.put(std::move(chargedHadronPUEnergyFractionV),"chargedHadronPUEnergyFraction");
-  
-  std::unique_ptr<edm::ValueMap<float>> chargedHadronCHSEnergyFractionV(new edm::ValueMap<float>());
-  edm::ValueMap<float>::Filler fillerCHE(*chargedHadronCHSEnergyFractionV);
-  fillerCHE.insert(srcJet,chargedHadronCHSEnergyFraction.begin(),chargedHadronCHSEnergyFraction.end());
-  fillerCHE.fill();
-  iEvent.put(std::move(chargedHadronCHSEnergyFractionV),"chargedHadronCHSEnergyFraction");
-
+  std::unique_ptr<edm::ValueMap<float>> valuesFPV0(new edm::ValueMap<float>());
```

Comment:
These chunks of copy/pasted code could also probably get compactified, although I have no elegant solution in mind for it.
If you don't have also, let it as such for now.

Reply:
I created a helper member function "fillValueMaps" for this purpose.